File: unslop/scripts/freshness/manifest.py
"""Concrete dependency manifest computation for ghost-staleness detection."""

from __future__ import annotations

from pathlib import Path

from ..core.frontmatter import parse_concrete_frontmatter
from ..core.hashing import MISSING_SENTINEL, UNREADABLE_SENTINEL, compute_hash
from ..dependencies.concrete_graph import get_all_strategy_providers


def _gather_recursive_providers(
    root: Path,
    meta: dict,
    seen: set | None = None,
) -> list[str]:
    """Recursively gather all strategy provider content from the full DAG.

    Walks concrete-dependencies and extends transitively, returning a
    sorted list of ``path:hash`` entries for every node in the tree.
    The ``seen`` set prevents infinite loops on circular references.
    """
    if seen is None:
        seen = set()

    entries = []
    providers = get_all_strategy_providers(meta)

    for dep_path in sorted(providers):
        if dep_path in seen:
            continue
        seen.add(dep_path)

        dep_full = root / dep_path
        if dep_full.exists():
            try:
                dep_content = dep_full.read_text(encoding="utf-8")
                entries.append(f"{dep_path}:{compute_hash(dep_content)}")
                # Recurse into this provider's own upstream
                dep_meta = parse_concrete_frontmatter(dep_content)
                entries.extend(_gather_recursive_providers(root, dep_meta, seen))
            except (OSError, UnicodeDecodeError):
                entries.append(f"{dep_path}:unreadable")
        else:
            entries.append(f"{dep_path}:missing")

    return entries


def compute_concrete_deps_hash(impl_path: str, project_root: str) -> str | None:
    """Compute a deep hash of all transitive strategy providers.

    Recursively walks concrete-dependencies and extends chains so that
    a change to a grandparent spec correctly invalidates all descendants.
    Returns a 12-char hex hash, or None if no strategy providers exist.
    Raises ValueError if the impl file exists but cannot be read.
    """
    impl = Path(impl_path)
    if not impl.exists():
        return None

    try:
        content = impl.read_text(encoding="utf-8")
    except (OSError, UnicodeDecodeError) as e:
        raise ValueError(f"Cannot read concrete spec: {impl} ({e})")

    meta = parse_concrete_frontmatter(content)
    providers = get_all_strategy_providers(meta)
    if not providers:
        return None

    root = Path(project_root).resolve()
    combined = _gather_recursive_providers(root, meta)
    return compute_hash("\n".join(combined))


def compute_concrete_manifest(impl_path: str, project_root: str) -> dict[str, str] | None:
    """Compute a per-dependency manifest for surgical ghost-staleness detection.

    Returns a dict of {dep_path: 12-char-hex-hash} for all **transitive**
    strategy providers (concrete-dependencies + extends parent, recursively).
    Returns None if no strategy providers exist.

    Unlike compute_concrete_deps_hash (which produces a single opaque hash),
    the manifest stores each dependency individually so check_freshness() can
    pinpoint exactly which dep changed -- including deep transitive changes.
    """
    impl = Path(impl_path)
    if not impl.exists():
        return None

    try:
        content = impl.read_text(encoding="utf-8")
    except (OSError, UnicodeDecodeError) as e:
        raise ValueError(f"Cannot read concrete spec: {impl} ({e})")

    meta = parse_concrete_frontmatter(content)
    direct_providers = get_all_strategy_providers(meta)
    if not direct_providers:
        return None

    root = Path(project_root).resolve()
    manifest = {}

    # BFS through the full transitive provider tree
    queue = list(direct_providers)
    visited: set[str] = set()

    while queue:
        dep_path = queue.pop(0)
        if dep_path in visited:
            continue
        visited.add(dep_path)

        dep_full = root / dep_path
        if dep_full.exists():
            try:
                dep_content = dep_full.read_text(encoding="utf-8")
                manifest[dep_path] = compute_hash(dep_content)
                # Walk this dep's own providers (transitive)
                dep_meta = parse_concrete_frontmatter(dep_content)
                for upstream in get_all_strategy_providers(dep_meta):
                    if upstream not in visited:
                        queue.append(upstream)
            except (OSError, UnicodeDecodeError):
                manifest[dep_path] = UNREADABLE_SENTINEL
        else:
            manifest[dep_path] = MISSING_SENTINEL

    return manifest if manifest else None


def format_manifest_header(manifest: dict[str, str]) -> str:
    """Format a concrete manifest dict as a header-safe string.

    Output: dep1.impl.md:a3f8c2e9b7d1,dep2.impl.md:7f2e1b8a9c04
    """
    return ",".join(f"{path}:{h}" for path, h in sorted(manifest.items()))

File: prunejuice/src/manifest.ts
import { readFile } from "node:fs/promises";
import { join, resolve } from "node:path";
import { truncatedHash } from "./hashchain.js";
import { isEnoent } from "./fs-utils.js";
import { parseConcreteSpecFrontmatter } from "./ripple.js";
import type { TruncatedHash, ManifestDiff } from "./types.js";
import { MISSING_SENTINEL, UNREADABLE_SENTINEL } from "./types.js";

// -- Manifest diffing (pure, no IO) ------------------------------------------

/** Compute the structural diff between two concrete dependency manifests. */
export function diffConcreteManifests(
  previous: Map<string, TruncatedHash>,
  current: Map<string, TruncatedHash>,
): ManifestDiff {
  const added: string[] = [];
  const removed: string[] = [];
  const changed: string[] = [];

  for (const [path, hash] of current) {
    const prev = previous.get(path);
    if (prev === undefined) {
      added.push(path);
    } else if (prev !== hash) {
      changed.push(path);
    }
  }

  for (const path of previous.keys()) {
    if (!current.has(path)) {
      removed.push(path);
    }
  }

  added.sort();
  removed.sort();
  changed.sort();

  return { added, removed, changed };
}

// -- Strategy provider resolution ---------------------------------------------

/** Get all strategy providers: concrete-dependencies union extends. */
function getAllStrategyProviders(meta: {
  concreteDependencies: string[];
  extends: string | null;
}): string[] {
  const providers = new Set(meta.concreteDependencies);
  if (meta.extends) providers.add(meta.extends);
  return [...providers].sort();
}

// -- Manifest computation (IO-bound BFS) --------------------------------------

/**
 * Compute per-dependency manifest for a concrete spec.
 * BFS through concrete-dependencies + extends, hashing each file's content.
 * Returns null if the spec has no providers.
 */
export async function computeConcreteManifest(
  specPath: string,
  cwd: string,
): Promise<Map<string, TruncatedHash> | null> {
  const absCwd = resolve(cwd);
  const absSpec = join(absCwd, specPath);

  let content: string;
  try {
    content = await readFile(absSpec, "utf-8");
  } catch (err) {
    if (isEnoent(err)) return null;
    throw err;
  }

  const meta = parseConcreteSpecFrontmatter(content);
  const directProviders = getAllStrategyProviders(meta);
  if (directProviders.length === 0) return null;

  const manifest = new Map<string, TruncatedHash>();
  const visited = new Set<string>();
  const queue = [...directProviders];

  while (queue.length > 0) {
    const depPath = queue.shift()!;
    if (visited.has(depPath)) continue;
    visited.add(depPath);

    const absDep = join(absCwd, depPath);
    let depContent: string;
    try {
      depContent = await readFile(absDep, "utf-8");
    } catch (err) {
      if (isEnoent(err)) {
        manifest.set(depPath, MISSING_SENTINEL);
        continue;
      }
      manifest.set(depPath, UNREADABLE_SENTINEL);
      continue;
    }

    manifest.set(depPath, truncatedHash(depContent));

    // Walk transitive providers
    const depMeta = parseConcreteSpecFrontmatter(depContent);
    for (const upstream of getAllStrategyProviders(depMeta)) {
      if (!visited.has(upstream)) {
        queue.push(upstream);
      }
    }
  }

  return manifest.size > 0 ? manifest : null;
}

/**
 * Compute a single opaque hash of all transitive strategy providers.
 * Returns null if the spec has no providers.
 */
export async function computeConcreteDepsHash(
  specPath: string,
  cwd: string,
): Promise<TruncatedHash | null> {
  const manifest = await computeConcreteManifest(specPath, cwd);
  if (!manifest) return null;

  const combined = [...manifest.entries()]
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([path, hash]) => `${path}:${hash}`)
    .join("\n");

  return truncatedHash(combined);
}
